### ISSUE TITLE
Create Db if it doesn't exist

### ISSUE THREAD

**Issue by llaslo** [DbUp/DbUp#74]
What I'd really like the is the ability to have DbUp have an option to create the database if it doesn't exist.  I tried to solve this by having my first script do this:

**0001_InitialCreate.sql**

    IF NOT EXISTS ( SELECT [Name] FROM sys.databases WHERE [name] = 'MyDatabase' )
    BEGIN
	    CREATE DATABASE [MyDatabase]
    END
    GO

    USE [MyDatabase]
    GO

    IF NOT EXISTS ( SELECT * FROM sysobjects where Name='MyTable' and xtype='U')
	    CREATE TABLE dbo.MyTable (
		    [Id] INT NOT NULL PRIMARY KEY IDENTITY,
		    [Name] VARCHAR(50) NOT NULL,
		    [CreatedBy] VARCHAR(50) NOT NULL,
		    [CreatedDate] DATETIME2(7) NOT NULL DEFAULT SYSDATETIME())
    GO

Subsequent scripts look like this:

**0002_AddAgeColumnToMyTable.sql**

    USE [MyDatabase]
    GO

    ALTER TABLE dbo.MyTable
    ADD [Age] INT NULL
    GO
    
This approach works fine if you leave the database name out of the connection string, and just use *USE* statements in your scripts.  The one problem, though, is that it doesn't find the SchemaVersions table so it re-runs every script every time.  (although it does actually find the table and insert the scripts into it).

**Comment by worldspawn** [comment]
```
using (var con = new SqlConnection(masterConnectionString))
            {
                using (var cmd = new SqlCommand())
                {
                    cmd.Connection = con;
                    con.Open();

                    cmd.CommandText = string.Format("SELECT COUNT([name]) FROM [sys].[databases] WHERE [name] = N'{0}'", database);
                    var databaseExists = Convert.ToBoolean(cmd.ExecuteScalar());
                    if (!databaseExists) CreateDatabase(con, database);
                    con.Close();
                }
}

public static void CreateDatabase(SqlConnection con, string database)
        {
            Console.Write("Creating database...");

            using (var cmd = new SqlCommand())
            {
                cmd.Connection = con;
                cmd.CommandText = string.Format("CREATE DATABASE {0}", database);
                cmd.ExecuteNonQuery();
            }

            Console.WriteLine("Done");
        }
```

**Comment by JakeGinnivan** [comment]
Happy to have a PR. It should be on the utils class, not in the upgrade API. That way if it is needed, you don't have to write the code but people new to DbUp don't need to understand the implications of just opting into creating the database as well

**Comment by aggieben** [comment]
@JakeGinnivan what do you mean by the "utils class"?  As far as I can tell, there's nothing with a name like that in the repo.

**Comment by JakeGinnivan** [comment]
We don't, but we could add one near https://github.com/DbUp/DbUp/blob/master/src/DbUp/Helpers/AdHocSqlRunner.cs?

**Comment by aggieben** [comment]
What about a new helper class called `Database` for db-level convenience methods?  There probably won't be a ton of those, but it doesn't quite feel like it belongs in `AdHocSqlRunner` to me.

**Comment by JakeGinnivan** [comment]
Yep agreed, maybe `DatabaseUtils`?

**Comment by aggieben** [comment]
Sure.  I'll give it a stab tonight and see what you think.

**Comment by JakeGinnivan** [comment]
Sweet, thanks!

**Comment by aggieben** [comment]
Actually, now that I'm looking at the Support classes, what about either an interface method (`IConnectionManager`) or an abstract method (in `DatabaseConnectionManager`) to do something like `.EnsureDb("name")`, and let each implementation do it's own thing?  They're all going to be a little different, and it seems like `IConnectionManager` and its implementors are the mechanism you've designed for db-specific support.

I'm pretty much brand-new to your code here and I'm just seeing some things for the first time ;-)

**Comment by JakeGinnivan** [comment]
That sounds reasonable to me, as long as it is an explicit call to enable the functionality. It is one of those features which people could run a web app under DBO just so dbup can create the database

**Comment by rickculpepper** [comment]
@aggieben I would like to have this functionality, but it does not appear to be in the master branch yet. Did you create a pull request for your changes? 

If so, @JakeGinnivan, is there a reason that this feature has not been merged to master?
Thanks!

**Comment by aggieben** [comment]
@rickculpepper yep, #99

**Comment by rickculpepper** [comment]
Thanks, Ben!

**Comment by rickculpepper** [comment]
The current NuGet package, version 3.2.1, was release in Nov 2014. So, I assume it does not have changes made since that time. Is there a plan to push a revised NuGet package anytime soon?

**Comment by JakeGinnivan** [comment]
On it, I have had this on my todo list for far too long. Am starting to work on a new release. I need to baseline the public API and check we have no SemVer breaks first (which I think we might, and why I stopped releasing). I needed to put the time aside for due diligence. While I am doing it I will move builds to AppVeyor and try and figure out if I can figure out a way to test DbUp without really hitting databases. 

I will aim to have a new release out, either 3.3.0 or v4 by next weekend.

**Comment by JakeGinnivan** [comment]
The PR only worked for Postgresql so this will not make the next release.

If someone else wants to pick it up please make it have an API similar to the deploy changes api. Maybe `CreateDatabase.For.SqlServer(...)` that way we can add extension methods to support each database type one by one. Thoughts?

**Comment by rickculpepper** [comment]
@JakeGinnivan, I am looking at this. I have some questions about your suggestion. I created a new class in DbUp like so.

```
using DbUp.Builder;

namespace DbUp
{
    /// <summary>
    /// A fluent builder for creating database upgraders.
    /// </summary>
    public static class CreateDatabase
    {
        private static readonly SupportedDatabases Instance = new SupportedDatabases();

        /// <summary>
        /// Returns the databases supported by DbUp.
        /// </summary>
        public static SupportedDatabases For
        {
            get { return Instance; }
        }
    }
}
```

That was easy enough. However, now I am looking at SqlServerExtensions.cs and think that we are going to have trouble distinguishing between extension methods for `DeployChanges.To` and `CreateDatabase.For`.

Both of these properties are the same type, `DbUp.Builder.SupportedDatabases`. So, the both sets of extension methods will have the same type for the `this` parameters. Since the signatures will be similar, if not identical, how do we distinguish them?

On a bit of reflection, there are two obvious ways to address this. 

1. The base class, `DeployChanges` or `CreateDatabase` could be used as the 'this' parameter of the extension, but this would be a breaking change for `DeployChanges.To`. 
2. `DeployChanges.To` and `CreateDatabase.For` could be of different types. Perhaps `DbUp.Builder.SupportedDatabases` should be in interface rather than an empty, concrete type. This might allow the extension methods to distinguish between `DeployChanges.To` and `CreateDatabases.For` without a breaking change. 

There are perhaps other techniques, but it I think they might require greater refactoring of the code base. 

I'm going to start down the path of #2 above and see how far I get. Looks like there are about 21 references of `DbUp.Builder.SupportDatabases` to be worked through.

I'll get back with you on my progress...

**Comment by JakeGinnivan** [comment]
`CreateDatabase.For` should have a new type it returns, it should have nothing to do with `DeployChanges.To`. `CreateDatabase.For` could return `CreateDatabaseFor` or any other type. Then just make the extension method on that.

**Comment by rickculpepper** [comment]
Okay. You said the same thing I was saying… just more succingtly… ;)

Thanks!

**Comment by JakeGinnivan** [comment]
Yeah, was pretty much repeating option 2 just to confirm :)

Sent from my Windows Phone
________________________________

**Comment by rickculpepper** [comment]
One other question. Looking at the Extensions methods, it appears that, apart from the `this` parameter and return value, they will have identical names and signatures. I know that is enough to distinguish them for the compiler. Is that enough to clearly distinguish them for we mere mortals? ;)

**Comment by JakeGinnivan** [comment]
The extension methods will only show for the correct type, so intellisense will only show the correct one.

**Comment by JakeGinnivan** [comment]
Unless you meant for the maintainers. :P in that case, might be a bit confusing. You could always keep them in different static classes to make it easier to tell

**Comment by dazinator** [comment]
Could you make the API express this as a "Create Database If Not Exists" operation rather than just a CreateDatabase - if you weren't going to do that already :)
Perhaps by changing the entry point of the fluent API from `CreateDatabase` to something like `EnsureDatabase` ? - not fussy as long as it ends up reading like it's an idempotent operation :)

**Comment by rickculpepper** [comment]
@JakeGinnivan, I have created pull request #114 to address this issue. I have implemented this as @dazinator requested above. 

Let me know if I'm going about this correctly. I have been in software development for 30+ years and have been developing with .NET since 2000 in the pre-release days. However, I am new to GitHub -- I'm a TFS VC junkie :+1 -- and this is my very first attempt to contribute to a project. I'm sure I screwed *something* up... ;)

**Comment by JakeGinnivan** [comment]
Fixed in #114 and released in 3.3.0! Thanks again @rickculpepper

**Comment by rickculpepper** [comment]
You are welcome, @JakeGinnivan!


Rick Culpepper
Senior Team Lead, Special Operations Team [SPOT]
Camping World | Good Sam
email: RCulpepper@CampingWorld.com<mailto:RCulpepper@CampingWorld.com>
office: (270) 901-4857<tel:(270)%20901-4857> | mobile: (615) 521-5763<tel:(615)%20521-5763> | internal : 79984857<tel:79984857>
websites: www.CampingWorld.com<http://www.campingworld.com/> |www.GoodSamClub.com<http://www.goodsamclub.com/> |www.CampingWorldRVSsales.com<http://www.campingworldrvssales.com/>

Sent from my iPhone

On Sep 16, 2015, at 6:34 PM, Jake Ginnivan <notifications@github.com<mailto:notifications@github.com>> wrote:


Fixed in #114<https://github.com/DbUp/DbUp/pull/114> and released in 3.3.0! Thanks again @rickculpepper<https://github.com/rickculpepper>

—
Reply to this email directly or view it on GitHub<https://github.com/DbUp/DbUp/issues/74#issuecomment-140925807>.

----------------------------------------------------------------------
NOTE:  The information in this email is confidential and may be legally 
privileged.  If you are not the intended recipient, we request that you 
(i) not read, use or disseminate the information, (ii) advise the sender 
immediately by reply email and (iii) delete this message and any 
attachments without retaining a copy.  Although this email and any 
attachments are believed to be free of any virus or other defect that 
may affect any computer system into which it is received and opened, it 
is the responsibility of the recipient to ensure that it is virus free 
and no responsibility is accepted by FreedomRoads, LLC or any of its 
affiliates for any loss or damage arising in any way from its use.
